cluster-autoscaler/simulator/framework/infos.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	apiv1 "k8s.io/api/core/v1"
+	resourceapi "k8s.io/api/resource/v1alpha3"
+	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+// PodInfo contains all necessary information about a Pod that Cluster Autoscaler needs to track.
+type PodInfo struct {
+	// This type embeds *apiv1.Pod to make the accesses easier - most of the code just needs to access the Pod.
+	*apiv1.Pod
+
+	// NeededResourceClaims contains ResourceClaim objects needed by the Pod.
+	NeededResourceClaims []*resourceapi.ResourceClaim
+}
+
+// NodeInfo contains all necessary information about a Node that Cluster Autoscaler needs to track.
+type NodeInfo struct {
+	// This type embeds the *schedulerframework.NodeInfo, and adds additional data on top - both on the
+	// Node and Pod level.
+	*schedulerframework.NodeInfo
+	// Pods should always match the embedded NodeInfo.Pods. This way the schedulerframework.PodInfo can be
+	// accessed via .NodeInfo.Pods[i], and the internal one via .Pods[i]. CA code normally only needs to
+	// access the Pod, so iterating over the internal version should be enough.
+	Pods []*PodInfo
```

Comment:
I understand why you need Pods with this extended information, but it's a bit scary that you're now having 2 sources of truth for the list of pods on a node (NodeInfo.Pods and NodeInfo.NodeInfo.Pods). It seems like this is asking for the two to somehow get out of sync. 

I'm not yet sure what the right solution is. Intuitively the correct way to handle this would be to have an internal representation with a single list of pods and convert on-demand, but I am worried about performance implications of that.

Reply:
Rather than embed a new, local `PodInfo` type can we simply embed `NeededResourceClaims []*resourceapi.ResourceClaim` directly? And follow `ResourceClaim.Status.ReservedFor` to get the associated pod references?

Reply:
@MaciekPytel fully agreed on this being scary, but the alternatives that I see are:
1. As you mention, changing the ClusterSnapshot implementations to operate only on the internal NodeInfos as the single source of truth, and converting them to `schedulerframework.NodeInfo` on-demand whenever scheduler code calls `NodeInfos().List()`. I went with this in the first prototype, but then decided against it for 2 reasons: 
  * As you point out, it's unclear how it would affect performance. This would have to be measured, but I suspect it might not be insignificant given that some things are precomputed during `schedulerframework.NodeInfo.AddPod()`.
  * It added unnecessary complexity to `ClusterSnapshot` implementations, and was annoyingly bug-prone. This is less important IMO, as at least the problems were limited to `simulator/snapshot`.
2. We could remove some redundancy by having the pod list only inside the wrapped `schedulerframework.NodeInfo`, and only keeping the additional information in `PodInfo`. So `PodInfo` would only contain `NeededResourceSlices`, and `NodeInfo` would contain `schedulerframework.NodeInfo` and a map from pod UIDs to `PodInfo`. We could also have a `Pods()` method on the internal NodeInfo that would merge the two data sources together for ease-of-use. This should be less prone to getting out-of-sync, but at a cost of more complexity. Do you think it's enough/worth it?

FWIW I tried to assert  in the unit test that things stay in sync during normal usage patterns that we have in the codebase right now.

@jackfrancis Unfortunately we can't rely on any information inside a ResourceClaim to identify the pod(s) that need it. `ReservedFor` is only filled out if the claim is actually allocated for the pod. But I think the spirit of your suggestion is similar to the option 2. I describe above, right?

Reply:
IMO that additional complexity is preferable to storing the same data in two places.

Reply:
Agree with @jackfrancis, duplicating data seems like the worst option. I think both approaches you listed could work. 

For approach 1 - you could opportunistically cache schedulerframework.NodeInfo to minimize performance impact (i.e. cache sf.NI whenever you build it, invalidate on any change to framework.NI). That still involves duplicated lists to an extent, but I think it's very different in that there is a clearly defined source of truth. It's probably doesn't solve all the performance problems, as you'd be building sf.NI each time and not updating it and IIRC you're correct and AddPod() was indeed optimized to efficiently update precomputes (for antiaffinity probalby?). I think this may be workable, but I don't love this option. 

Approach 2 seems better to me. As long as we can come up with a good interface that encapsulates the complexity (e.g. the Pods() method you described), I think it's a good option. The main thing is to make sure users of framework.NodeInfo won't have to know the information is stored separately (I'd go as far as making fields private and only allowing access via methods outside of simulator).

I think the best approach would be to extend NodeInfo in scheduler framework (i.e. make change to kubernetes/kubernetes).
I think the fact that scheduler snapshots nodes and pods, but just uses in-plugin informers for everything else is inconsistent, error prone and all around terrible. Including DRA objects in scheduler SharedLister seems like a step in the right direction from perspective of just the scheduler. And it has the added benefit of completely removing the need for any custom wrappers in CA together with all the complexity they bring. 
Unfortunately, I'm guessing we're out of time for such change in 1.32 scheduler. Even more unfortunate is that once we implement the wrappers, I can't see anyone going back and cleaning this up for 1.33 :(

So, to sum up - I'm +1 for approach 2.

Reply:
Is there a case to be made that extending NodeInfo in scheduler could have general reuse value? If so, we could probably find someone to do that work (I'd be happy to give that a shot).

This is just to say IMO approach 2 is the best overall, and we may still have a chance at an even better long term outcome if we have the appetite.

Reply:
@MaciekPytel Done, went with approach 2. Also made the fields private as you suggested. WDYT?

IMO having an internal NodeInfo type wrapping the schedulerframework one is useful for CA outside of the DRA logic. Not all info that CA would want to attach to a Node is relevant to scheduler, and putting it there just for CA would be confusing. E.g. now that we have this type, we could keep information like whether a Node is upcoming properly in there instead of hacking it inside Node annotations. Same for e.g. the boot-disk info that GCE cloud provider passes through annotations. The added complexity, on the other hand, should be hidden from most of the codebase.

Separately, I agree that the ideal solution would be for the scheduler to make all API objects that plugins might need a part of its Snapshot - not just Nodes and Pods. I explored that solution, but got to a conclusion that it'd be a massive change, and maybe it wouldn't even solve the problem fully:
1. The Node/Pod cache that feeds the scheduler Snapshot is quite complex. If we wanted to correlate the DRA objects to Nodes/Pods on the fly during informer updates, we'd need to be very careful not to bloat the complexity further.
2. Not all DRA objects are "bound" to a Node or a Pod. So we'd need to expand the scheduler Snapshot beyond the current NodeInfo/PodInfo anyway to track those. At this point, maybe it makes more sense not to correlate the objects on the fly, and just essentially have an additional snapshot of DRA objects alongside NodeInfos. In which case we'd need extra fields just in NodeInfo just for CA anyway.
3. The plugins would still need access to live informers to handle queuing hints. Not a problem for CA since CA doesn't run the queuing logic, but it's a readability challenge on the scheduler side.

@jackfrancis IMO extending the NodeInfo on scheduler side without doing the huge scheduler effort described above doesn't make sense, as the scheduler plugin wouldn't use these fields anyway - they would only be used by CA.